testgen/src/tester.rs

```diff
@@ -140,60 +263,79 @@ impl Tester {
     where
         T: 'static + DeserializeOwned + UnwindSafe,
     {
-        let test_env = self.env();
-        let test_fn = move |input: &str| match parse_as::<T>(&input) {
-            Ok(test_case) => Tester::capture_test(test_env.clone(), |env| {
+        let test_fn = move |_path: &str, input: &str| match parse_as::<T>(&input) {
+            Ok(test_case) => Tester::capture_test(|| {
                 test(test_case);
-                env
             }),
             Err(_) => ParseError,
         };
-        self.tests.push((name.to_string(), Box::new(test_fn)));
+        self.tests.push(Test {
+            name: name.to_string(),
+            test: Box::new(test_fn),
+        });
     }
 
-    pub fn add_test_with_env<T>(&mut self, name: &str, test: fn(T, &mut TestEnv))
+    pub fn add_test_with_env<T>(&mut self, name: &str, test: fn(T, &TestEnv, &TestEnv, &TestEnv))
```

Comment:
Why the `fn(...)` type for `test` instead of `impl FnOnce(...)`? IIRC the former denotes a function pointer and would thus not allow passing a closure here.

Reply:
Well, I've struggled a lot experimenting with different `FnOnce`, `Box<dyn FnOnce>`, etc. you may try yourself and follow the compiler errors;) The problem is that I need to store a function to call it later in `panic::catch_unwind`. This works fine with function pointers, but storing the closure is not that simple -- at least I have not managed to find a way to store a closure, and at the same time to satisfy the requirement that it can be used across unwind boundaries.

Reply:
I see. Perhaps if you require a `Fn() + UnwindSafe + RefUnwindSafe`? Seems to work in [this simplified version](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b7643e64b74816d1b6841bd9312cb841) (which hopefully reproduces the same requirements).

Reply:
Wow! THIS indeed works without any problems:)) The peace I was missing is `RefUnwindSafe`. Thanks a ton!